pallets/phala/src/registry.rs

```diff
@@ -337,6 +372,57 @@ pub mod pallet {
 			Ok(())
 		}
 
+		#[pallet::weight(0)]
+		pub fn upload_code(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResult {
+			ensure_signed(origin)?;
+			let code_hash = T::Hashing::hash(&code.encode());
+			PristineCode::<T>::insert(&code_hash, &code);
+			Self::deposit_event(Event::CodeUploaded(code_hash));
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn instantiate_contract(
+			origin: OriginFor<T>,
+			// #[pallet::compact] endowment: BalanceOf<T>,
+			// #[pallet::compact] gas_limit: Weight,
+			code_index: CodeIndex<CodeHash<T>>,
+			data: Vec<u8>,
+			salt: Vec<u8>,
+			deploy_worker: Option<WorkerPublicKey>,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+
+			match code_index {
+				CodeIndex::NativeCode(_) => {}
+				CodeIndex::WasmCode(code_hash) => {
+					ensure!(
+						PristineCode::<T>::contains_key(code_hash),
+						Error::<T>::CodeNotFound
+					);
+				}
+			}
+
+			let group_counter = ContractGroupCounter::<T>::mutate(|group_counter| {
+				*group_counter += 1;
+				*group_counter
+			});
+			// we send hash instead of raw code here to reduce message size
+			let contract_info = ContractInfo {
+				owner,
+				group_counter,
+				salt,
```

Comment:
Ah, the contract address = hash(deploying_address + code_hash + salt).
We must find a way to avoid address collision in different groups, espacially for the contract instantiated contratcts.

Reply:
Maybe we should let pink_address = hash(group_id, ink_address)? @h4x3rotab

Reply:
To me it's better to consider the group id when generating the ink address. In this way we only have the ink address, and it should be globally unique